Serial Component

barreeeiroo opened the pull request:
This PR brings the current Arduino component at Kodular to MIT App Inventor 2. We think this can be really useful to have at AI, as lots of schools have Arduino kits. Having a native Arduino component at AI will bring several new possibilities to them.
It is based on [this open-sourced extension](https://gist.github.com/pavi2410/d6a9e11a99421dff430b394579e6834d) from @pavi2410, but we made some modifications to it. This component is exactly the same as the one at Kodular.

NOTE: This PR will conflict with #1731 once merged (and vice-versa). When any of these PRs gets merged, the other one needs to have some conflicts resolved. Just ping me and I'll fix it.

AppInventorWorkerBee commented:
Can one of the admins verify this patch?

barreeeiroo commented:
@ewpatton will keep working on this next week to solve those issues 👍
Regarding the documentation, we can provide the following guide: http://extensiondirectory.com/tools/arduino-extension.html

1. Not sure to be honest. @pavi2410 can provide more info on this
2. Yeah, I remember talking with someone on the Summit and suggesting the same thing. We will consider it

Sanjit1 commented:
> @barreeeiroo Two more things that would be helpful for us to know:
> 
> 1. What hardware does this need to connect up the Arduino with the phone?
> 2. It seems like it is just a serial port library, so in theory it should work with any device that can do serial communication (in addition to the Arduino). Is that correct? If so, we may want to consider renaming it to `Serial` or `SerialPort`. 


 - When I used it, I connected it to the USB micro of my phone. But I could not get it to work on my dad's phone, because it did not have OTG support. 
 - This will be a valid change in my opinion because I used it to connect to multiple USB components

barreeeiroo commented:
@ewpatton just updated this pull request, applying your changes:
1. Fixed indentation.
2. Fixed import wildcards.
3. Renamed component to Serial.
_You were right, it not only for Arduino devices; it's just a generic serial connector. However, I preferred to leave the Arduino icon as it is a "standard" for that connection, and users will mostly use it with that._
4. Added documentation.

ewpatton commented on appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java:
```diff
@@ -5181,4 +5181,45 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
       "<br/>These applications will not run on Android versions older than 4.0.")
   @Description("Text for the Package SDK 26 Warning Dialog Box (HTML)")
   String Package26Notice();
+
```
These should all be handled by AutogeneratedOdeMessages now. Do you still need them here?

barreeeiroo replied:
I added them because of habit. In Kodular, we still use OdeMessages (as we like using spaces in names), so I added them here just in case.

ewpatton commented on appinventor/components/build.xml:
```diff
@@ -152,6 +152,7 @@
     <copy toFile="${public.deps.dir}/commons-pool.jar" file="${lib.dir}/commons-pool/commons-pool2-2.0.jar" />
     <copy toFile="${public.deps.dir}/guava-14.0.1.jar" file="${lib.dir}/guava/guava-14.0.1.jar" />
     <copy toFile="${public.deps.dir}/core.jar" file="${lib.dir}/QRGenerator/core.jar" />
+    <copy toFile="${public.deps.dir}/physicaloid.jar" file="${lib.dir}/physicaloid/physicaloid-library.jar" />
```
What's the min SDK needed for this lib? If it's not minSdk 7, please specify in the component.

barreeeiroo replied:
Sorry, totally forgot about it. As we have a higher min sdk I didn't add it here. It is Android 3.1 = API 12.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/common/YaVersion.java:
```diff
@@ -488,8 +488,10 @@ private YaVersion() {
   // - PEDOMETER_COMPONENT_VERSION was incremented to 2
   // For YOUNG_ANDROID_VERSION 196:
   // - WEBVIEWER_COMPONENT_VERSION was incremented to 8
+  // For YOUNG_ANDROID_VERSION 197:
```
Generally I would recommend not bumping YOUNG_ANDROID_VERSION so long as you mark the PR editable by maintainers. This allows us to push a commit to your branch when we're ready to merge since we don't necessarily guarantee we can merge them in a particular order. Since this seems the most likely candidate right now I'll let it slide.

barreeeiroo replied:
Fixed, reverted these changes.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+
+  private Context context;
+
+  Physicaloid mPhysicaloid;
```
This can be made private.

barreeeiroo replied:
Fixed.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+
+  @SimpleFunction(description = "Initializes serial connection.")
+  public void InitializeSerial() {
+    mPhysicaloid = new Physicaloid(context);
```
Might make sense to call `setBaudrate` here just in case the default changes in a future version.

barreeeiroo replied:
Done.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+  @SimpleFunction(description = "Opens serial connection. Returns true when opened.")
+  public boolean OpenSerial() {
+    Log.d(LOG_TAG, "Opening connection");
+    return mPhysicaloid.open();
```
Potential NPE if the blocks aren't called in the right order. This is true in most other functions in this class.

barreeeiroo replied:
Solved, as well as in other cases.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+
+  @SimpleFunction(description = "Sets a new baud rate. Default is 9600 bps.")
+  public void BaudRate(int baudRate) {
+    this.baudRate = baudRate;
```
What's the purpose of storing this if it's never read? I'm curious why you chose to make a function rather than a property with setter/getter.

barreeeiroo replied:
Because it's actually a function in the library. It is not something you can "predefine". And, well, it is how we had it in Kodular. Now I realize it could be improved a bit more. I've now changed it to a getter-setter property. 😓

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+
+  @SimpleFunction(description = "Reads data from serial.")
+  public void ReadSerial() {
+    byte[] buf = new byte[256];
```
Is the data in the buffer always guaranteed to be < 256 bytes? What happens if you're in the middle of reading a multibyte UTF-8 character? It seems like you'd possibly lose data. Is the `read()` call blocking?

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+  public void WriteSerial(String writeDataSerial) {
+    if (!writeDataSerial.isEmpty()) {
+      byte[] buf = writeDataSerial.getBytes();
+      mPhysicaloid.write(buf);
```
Since it looks like `write` can return -1 in the case of an error, it may make sense to test for this and report the problem either through an error event on the component or via the screen's ErrorOccurred event.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  @SimpleFunction(description = "Writes given data to serial.")
+  public void WriteSerial(String writeDataSerial) {
```
I wonder if it might also make sense to have a "PrintSerial" function that automatically includes the \n.

barreeeiroo replied:
Added new method. However, I think that, if the user wants to add "\n" he will add it in blocks. We may get questions about why having both blocks.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+    }
+  }
+
+  @SimpleFunction(description = "Returns true when the Serial connection is open.")
```
Consider making it a read-only property "Connected" or "Opened"

barreeeiroo replied:
Changed to "Opened" property.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,109 @@
+
+    if (mPhysicaloid.read(buf) > 0) {
+      try {
+        data = new String(buf, "UTF-8");
```
Since `read` looks to be synchronous, it might make sense for this to return the data (or empty string) rather than reporting through an event. Otherwise, what happens is the running of the event preempts the execution of the current blocks stack where this function is executing, which is not in line with the existing event model. Alternatively, post a runnable with the event callback to the UI handler rather than immediately dispatching below.

ewpatton commented:
@barreeeiroo You should also include the changes to the markdown documentation files in addition to the HTML ones. If you've got Jekyll and Minima installed on your machine, they should run automatically as part of the build process (assuming you have the latest master/ucr).

barreeeiroo commented:
> @barreeeiroo You should also include the changes to the markdown documentation files in addition to the HTML ones. If you've got Jekyll and Minima installed on your machine, they should run automatically as part of the build process (assuming you have the latest master/ucr).

I've applied some of the changes. There were a few comments that I could not yet check but will do this weekend. Also, I still haven't updated documentation (will do ASAP). I will let you know once ready.

barreeeiroo commented:
Alright, so I think I have addressed all the issues with the components applying all your comments.
But I was not sure about the 256 bytes buffer size. With a lack of official documentation and usage examples, I was not able to see if it will work for all cases. Thus, I preferred setting it as Designer Property and let the user decide if modify it or not, so it will be 256 by default. Although, in theory it should work with any size.

ewpatton commented on appinventor/components/src/com/google/appinventor/components/runtime/Serial.java:
```diff
@@ -0,0 +1,164 @@
+          Log.e(LOG_TAG, mEr.getMessage());
+        }
+      } else {
+        success = false;
```
Success is never used.
